android/app/src/main/kotlin/io/lantern/model/SessionModel.kt

```diff
@@ -396,28 +401,53 @@ class SessionModel internal constructor(
                 val start = System.currentTimeMillis()
                 val appsList = appsDataProvider.listOfApps()
                 // First add just the app names to get a list quickly
-                val apps = buildJsonArray {
-                    appsList.forEach { app ->
-                        add(
-                            buildJsonObject {
-                                val byte = app.icon!!
-                                put("packageName", app.packageName)
-                                put("name", app.name)
-                                putJsonArray("icon") {
-                                    byte.toUByteArray().forEach { add(it.toInt()) }
+                val file = File(activity.cacheDir, "appsData.json")
+
+
+                // Write JSON data directly to the file to reduce memory usage
+                file.outputStream().use { outputStream ->
+                    Json.encodeToStream(buildJsonArray {
+                        appsList.forEach { app ->
+                            add(
+                                buildJsonObject {
+                                    val byte = app.icon ?: ByteArray(0)  // Safeguard for null icons
+                                    put("packageName", app.packageName)
+                                    put("name", app.name)
+                                    putJsonArray("icon") {
+                                        byte.toUByteArray().forEach { add(it.toInt()) }  // Store each UByte as an Int
+                                    }
                                 }
-                            }
-                        )
-                    }
+                            )
+                        }
+                    }, outputStream)
                 }
```

Comment:
This looks like a great way to mitigate some of the memory issues! I think we could optimize further by implementing some caching here as well? For instance, you could serve the JSON file from cache if it already exists and add logic to refresh it only if it’s older than 24 hours. Additionally, you could minimize file writes by re-generating the file only if the installed apps differs from the last snapshot? We could always handle this in a subsequent PR but I figured I would mention it!

Reply:
Thanks, Looking into it.

Reply:
Changes are done.